node/core/av-store/src/lib.rs

```diff
@@ -110,62 +348,201 @@ impl AvailabilityStoreSubsystem {
 			metrics: Metrics(None),
 		}
 	}
 }
 
-async fn run<Context>(subsystem: AvailabilityStoreSubsystem, mut ctx: Context)
+fn get_next_pov_pruning_time(db: &Arc<dyn KeyValueDB>) -> Option<NextPoVPruning> {
+	query_inner(db, columns::META, &NEXT_POV_PRUNING)
+}
+
+fn get_next_chunk_pruning_time(db: &Arc<dyn KeyValueDB>) -> Option<NextChunkPruning> {
+	query_inner(db, columns::META, &NEXT_CHUNK_PRUNING)
+}
+
+async fn run<Context>(mut subsystem: AvailabilityStoreSubsystem, mut ctx: Context)
 	-> Result<(), Error>
 where
 	Context: SubsystemContext<Message=AvailabilityStoreMessage>,
 {
 	let ctx = &mut ctx;
 	loop {
+		let pov_pruning_time = subsystem.maybe_prune_povs()?;
+		let chunk_pruning_time = subsystem.maybe_prune_chunks()?;
+
+		let mut pov_pruning_time = pov_pruning_time.fuse();
+		let mut chunk_pruning_time = chunk_pruning_time.fuse();
+
 		select! {
 			incoming = ctx.recv().fuse() => {
 				match incoming {
-					Ok(FromOverseer::Signal(Conclude)) => break,
-					Ok(FromOverseer::Signal(_)) => (),
+					Ok(FromOverseer::Signal(OverseerSignal::Conclude)) => break,
+					Ok(FromOverseer::Signal(OverseerSignal::ActiveLeaves(
+						ActiveLeavesUpdate { activated, .. })
+					)) => {
+						for activated in activated.into_iter() {
+							process_block_activated(ctx, &subsystem.inner, activated).await?;
+						}
+					}
+					Ok(FromOverseer::Signal(OverseerSignal::BlockFinalized(hash))) => {
+						process_block_finalized(&subsystem, ctx, &subsystem.inner, hash).await?;
+					}
 					Ok(FromOverseer::Communication { msg }) => {
-						process_message(&subsystem.inner, &subsystem.metrics, msg)?;
+						process_message(&mut subsystem, ctx, msg).await?;
 					}
 					Err(_) => break,
 				}
 			}
+			pov_pruning_time = pov_pruning_time => {
+				subsystem.prune_povs()?;
+			}
+			chunk_pruning_time = chunk_pruning_time => {
+				subsystem.prune_chunks()?;
+			}
 			complete => break,
 		}
 	}
 
 	Ok(())
 }
 
-fn process_message(db: &Arc<dyn KeyValueDB>, metrics: &Metrics, msg: AvailabilityStoreMessage) -> Result<(), Error> {
+/// As soon as certain block is finalized its pruning records and records of all
+/// blocks that we keep that are `older` than the block in question have to be updated.
+///
+/// The state of data has to be changed from
+/// `CandidateState::Included` to `CandidateState::Finalized` and their pruning times have
+/// to be updated to `now` + keep_finalized_{block, chunk}_for`.
+async fn process_block_finalized<Context>(
+	subsystem: &AvailabilityStoreSubsystem,
+	ctx: &mut Context,
+	db: &Arc<dyn KeyValueDB>,
+	hash: Hash,
+) -> Result<(), Error>
+where
+	Context: SubsystemContext<Message=AvailabilityStoreMessage>
+{
+	let block_number = get_block_number(ctx, hash).await?;
+
+	if let Some(mut pov_pruning) = pov_pruning(db) {
+		// Since the records are sorted by time in which they need to be pruned and not by block
+		// numbers we have to iterate through the whole collection here.
+		for record in pov_pruning.iter_mut() {
```

Comment:
How large can the `pov_pruning` be? I guess blocks are not finalized that often, but executing a CPU-heavy task in an async context is not nice. We could use a different datastructure to reduce latencies here.

Reply:
so around 20k records (we generate as much in a day if I am not wrong and we keep stuff for 1 day). I agree that this needs to be optimised but it boils down to the problem that we will inevitably hit the linear time here (change my mind)

Reply:
In any case there will be at least linear time when the array is lifted and decoded ?

Reply:
Speaking of, it doesn't seem like a good idea to put and read a 20k vec every single time especially on block activated?
A different db layout would be preferred, I'll think about alternatives.

Reply:
On every activated block we are going to be writing/reading `AvailabilityData` or `ErasureChunk`s which are expected to be substantially larger in size. If the we keep 20k records we are looking at ~100kb record array. With that I would suggest approaching optimizing the layout after benchmarking